### DISCUSSION TITLE
weird behavior in API routes when using setInterval function

### DISCUSSION THREAD

**Discussion by amir-khoshbakht** [vercel/next.js#40442] [Help]
I just noticed that interval timer doesn't stop when nextjs detects change in code and recompiles the project. is this normal ?

steps to reproduce :
1-pages/api/hello.ts :
```
export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<Data>
) {   
  const randomNumber = Math.floor(Math.random() * 10000000);

  setInterval(() => {
    console.log(randomNumber);
  }, 1000);

  res.status(200).end();
}
```
2-open this in browser :
 [http://localhost:3000/api/hello](http://localhost:3000/api/hello)

**Comment by icyJoseph** [comment]
> Caveat: I tried a bunch of things while writing this, so it might actually make very little sense 😅 

I guess it is expected? It is a leaking handler after all. AFAIK, when developing it is the same node process that runs this asynchronous work, so compiling a new handler shouldn't imply cleaning up tasks initiated by the old code.

In a serverless environment, this might prevent the runner from closing the handler, potentially running into timeout failures, or plain old memory leaks.

One way to look at it is that, the code is compiled, and then a request arrives, which runs that code, and the task it creates never ends because of the interval.

Then another request arrive, the code runs again, and a new tasks spins up. Changing the code just changes what the next incoming request will use, but there's no hooks to tell the already running code that it has to terminate the interval and stop. 

Again in a production environment, serverless functions are already compiled, and they are executed as tasks, with a timeout, so even if there's nothing to tell the function itself to exit the interval, the task runner will exit it sooner or later.

One particularly annoying thing is that the compilation step doesn't run any code, even if at the top level of the module, until a request comes through. 

I had thought that from `next.config.js` we could setup a WeakMap and keep tabs of the running version of `api/hello`, but it only works if a new request arrives, otherwise, the old ones keep on running.

**Comment by sti-web** [comment]
Hi @amir-khoshbakht,
I have the same,
If I create an interval with setInterval, the content of this component get re-rendered each interval cycle.
```
useEffect(() => {
    let progressInterval = setInterval(() => {
       // ...
    }, 800);

    return () => {
      if (progressInterval) {
        clearInterval(progressInterval);
      }
    };
  }, []);
```
I'm trying to find another way to have a component with timers

**Reply by icyJoseph** [reply]
What do you do inside the timer callback?

**Reply by sti-web** [reply]
Basically, if you connect interval body function with the component state, then whole component will be re-rendered. This is known feature of React.
I want to re-render only 1 single component on the page,` <ProgressBar /> `
I tried to use useRef(), but the latest value won't be populated to component. like `<ProgressBar value={ref.current} /> `
The same with useContext,
I think, I will try redux and state, with useReducer.
Like described here: https://itnext.io/how-to-work-with-intervals-in-react-hooks-f29892d650f2

**Reply by icyJoseph** [reply]
Right, if you call a state setter, you'll trigger a re-render. 

The default React behaviour is to re-render the entire tree down from where the state change happens. Also called standard behaviour.

- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#standard-render-behavior

> This means that calling setState() in your root <App> component, with no other changes altering the behavior, will cause React to re-render every single component in the component tree. After all, one of the original sales pitches for React was ["act like we're redrawing the entire app on every update"](https://www.slideshare.net/floydophone/react-preso-v2).

What you can do is use render props. 


```tsx
const TimerInjector = ({children}) => {
  const [data, setData] = useState(null)
  useEffect(() => {
    let progressInterval = setInterval(() => {
       setData(/* something something */)
    }, 800);

    return () => {
      if (progressInterval) {
        clearInterval(progressInterval);
      }
    };
  }, []);
  
  return children({data});
} 
```

And then you use it like so:

```tsx
const View = () => {

  return (
    <>
     <div>foo</div>
     <TimerInjector>
       {({data}) => <ProgressBar value={data} /> }
     </TimerInjector>
    </>
   )
}
```

And now you've isolated the update to that section of the JSX only, while still keeping TimerInjector open to take data from View and change its logic based off that.

A running example: https://codesandbox.io/s/staging-dust-28qz6x?file=/src/App.js

**Reply by sti-web** [reply]
@icyJoseph thanks. I'll try it tomorrow.
I'm curious about forwardRef hook, I could pass ref, and update children by that reference.

**Reply by sti-web** [reply]
@icyJoseph Thank you very much.
TimerInjector works well.
This is final code:
```
  const TimerInjector = ({ children }) => {
    const [count, setCount] = useState(0);
    useEffect(() => {
      let progressInterval = setInterval(() => {
        setCount((count) => count + 1);
      }, 1000);

      return () => {
        if (progressInterval) {
          clearInterval(progressInterval);
        }
      };
    }, []);

    return children({ count });
  };
...
  <TimerInjector>
    {({ count }) => <ProgressBar progress={count} />}
  </TimerInjector>
```
